compiler/hash-ir/src/ir.rs

```diff
@@ -1,59 +1,71 @@
 //! Hash Compiler Intermediate Representation (IR) crate. This module is still
 //! under construction and is subject to change.
+#![allow(unused)]
 
-use hash_source::{identifier::Identifier, location::Span};
-use hash_utils::counter;
-
-counter! {
-    name: IrId,
-    counter_name: IR_ID_COUNTER,
-    visibility: pub,
-    method_visibility: pub,
-}
-
-// TODO: do we need namespaces, we could just de-sugar them into binds that are
-// associated       with symbols?
-
-// TODO: We could just de-sugar guard patterns into:
-//
-// From:
-//
-// ```
-// match k {
-//    1 if x => ...
-//}
-// ```
-// 
-// Into:
-// ```
-// match k {
-//     1 => if x {
-//         ...
-//     }
-// }
-// ```
-// 
-// But, this means that we have to perform exhaustiveness checking before transforming into IR?
+use hash_source::{
+    constant::{InternedFloat, InternedInt, InternedStr},
+    location::{SourceLocation, Span},
+    SourceId,
+};
+use index_vec::{index_vec, IndexSlice, IndexVec};
 
+/// Represents the type layout of a given expression.
 #[derive(Debug, PartialEq, Eq)]
-pub enum PatKind<'i> {
-    Spread,
-    Wild,
-    Bind(Identifier),
-    Lit(Const),
-    Tuple(&'i [Pat<'i>]),
-    Constructor(&'i Pat<'i>, &'i [Pat<'i>]),
-    List(&'i [Pat<'i>]),
-    Union(&'i [Pat<'i>]),
-}
-
-/// A Pattern within IR
-#[derive(Debug, PartialEq, Eq)]
-pub struct Pat<'i> {
-    /// The kind of the pattern
-    kind: &'i PatKind<'i>,
-    /// The span of the pattern
-    span: Span,
+pub enum Ty<'i> {
+    /// `usize` type, machine specific unsigned pointer
+    USize,
+    /// `u8` type, 8bit unsigned integer
+    U8,
+    /// `u16` type, 16bit unsigned integer
+    U16,
+    /// `u32` type, 32bit unsigned integer
+    U32,
+    /// `u64` type, 64bit unsigned integer
+    U64,
+    /// `isize` type, machine specific unsigned pointer
+    ISize,
+    /// `i8` type, 8bit signed integer
+    I8,
+    /// `i16` type, 16bit signed integer
+    I16,
+    /// `i32` type, 32bit signed integer
+    I32,
+    /// `i64` type, 64bit signed integer
+    I64,
+    /// `f32` type, 32bit float
+    F32,
+    /// `f64` type, 64bit float
+    F64,
+    /// `char` type, 32bit characters
+    Char,
+    /// A `void` type
+    Void,
+    /// Represents any collection of types in a specific order.
+    Structural(&'i [Ty<'i>]),
+    /// Essentially an enum representation
+    Union(&'i [Ty<'i>]),
+    /// Pointer that points to some type
+    Ptr(&'i Ty<'i>),
+    /// Raw Pointer that points to some type
+    RawPtr(&'i Ty<'i>),
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Const {
+    /// Character constant
+    Char(char),
+    /// Integer constant that is defined within the program source.
+    Int(InternedInt),
+    /// Float constant that is defined within the program source.
+    Float(InternedFloat),
+
+    /// String has to become a struct that has a pointer
```

Comment:
Maybe this is a misleading comment? These are static strings no?

Reply:
Yeah, well all of the strings here will be static, everything else is a pointer